README.md
# Hiro's Quick Links

### A Chrome browser extension

#### Features

* Populates a panel with handy WooCommerce navigation links
* Toggle WordPress notices (for screenshots, etc)
* Automatically hides any Yoast SEO columns added in the products page

#### Usage notes:

When first signing in to a WooCommerce site, the links will take a few seconds to several seconds to load (be patient). After the links have been loaded once, they are retained for the life of the browser session.

#### Installation

##### From the repo

1. Clone the [repo](https://github.com/WillBrubaker/hiro-quick-links)
2. See the **Load the extension** heading here: [Getting Started: Building a Chrome Extension](https://developer.chrome.com/extensions/getstarted#unpacked)

##### Packaged extension file

1. ~~[click here (or right click - save as)](https://github.com/WillBrubaker/hiro-quick-links/blob/master/hiro-quick-links.crx?raw=true)~~
2. ~~That will download hiro-quick-links.crx to you default Downloads directory. It will also give you a notice that extensions can not be installed from this website.~~
3. ~~In your chrome location bar chrome://extensions/~~
4. ~~Drag & drop hiro-quick-links.crx from your downloads folder to the browser window~~

Chrome no longer allows packaged extensions unless they come from the Chrome web store.


##### License

This extension is released under the terms of the [GPLv2](http://www.gnu.org/licenses/gpl-2.0.html) It does contain components that have their own separate license terms.

jQuery is licensed under the [MIT license](https://jquery.org/license/)

qTip is licensed under the [MIT license](https://github.com/qTip2/qTip2#license)

##### Bug reports and code contributions

Bug reports and pull requests can be submitted on the [github repo](https://github.com/WillBrubaker/hiro-quick-links)

CHANGELOG.md
#### v0.7
* WooCommerce 2.6 compatibility added
* "Add to cart" links added to the products screen (expreimental)
* Hide wp-pointers if "hide yoast" is enabled.
* Tooltip positioning improvement
* adds changelog

#### v0.6

* Fix for frontend page links

#### v0.5

* Add GPL compatible tooltip library

content.js
/**
 * License: GPLv2 http://www.gnu.org/licenses/gpl-2.0.html
 * Copyright (C) 2015  Will Brubaker
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
var siteContent
var site
var protocol
var underLinks = []
var noticesDisplayed = true
var wcPages = {
	"Cart": null,
	"Checkout": null,
	"My Account": null,
	"Shop": null
}

jQuery(document).ready(function() {
	/*window variables aren't accessible to extension content scripts, need to get the ajax url somehow*/
	var ajaxUrl = retrieveAjaxUrl()
	if ("undefined" != typeof ajaxUrl) { //ajaxurl is defined, this is probably a WordPress site
		/*try to get a base url for the site*/
		var link = document.createElement("a")
		link.href = ajaxUrl
		site = link.host + link.pathname.replace("/wp-admin/admin-ajax.php", "")
		protocol = link.protocol
		message = {
			"action": "have-content",
			"value": {
				"site": site
			}
		}
		chrome.runtime.sendMessage(message, function(response) { //ask the background script for content for this site
			if ("undefined" != response.content && false == response.content) { //bg script didn't have content - fetch new
				jQuery.when(jQuery.get(protocol + "//" + site + "/wp-admin/admin.php?page=wc-settings", function(data) {
					links = jQuery('h2.woo-nav-tab-wrapper a, nav.woo-nav-tab-wrapper a', data)
					wcLinks = '<ul>'
					jQuery(links).each(function(index) {
						anchorText = jQuery(this).html()
						href = jQuery(this).attr('href')
						if (index > 0) {
							wcLinks += '<li>'
						}
						wcLinks += '<a href="' + href + '" tabindex="-1">' + anchorText + '</a>'
						if (index > 0) {
							wcLinks += '</li>'
						}
					})
					wcLinks += '</ul>'
				}).fail(function(xhr) {
					console.log(xhr)
				}).done(function() {
					jQuery.when(jQuery.get(protocol + "//" + site +
						"/wp-admin/admin.php?page=wc-settings&tab=products&section=display",
						function(data) {
							shopPageId = jQuery("#woocommerce_shop_page_id option:selected", data).val()
							if ("undefined" != typeof shopPageId && shopPageId.length > 0) {
								wcPages.Shop = protocol + "//" + site + "/?p=" + shopPageId
							}
						}).done(function() {}))
					jQuery.when(jQuery.get(protocol + "//" + site + "/wp-admin/admin.php?page=wc-settings&tab=checkout",
						function(data) {
							cartPageId = jQuery("#woocommerce_cart_page_id option:selected", data).val()
							if ("undefined" != typeof cartPageId && cartPageId.length > 0) {
								wcPages.Cart = protocol + "//" + site + "/?p=" + cartPageId
							}
							checkoutPageId = jQuery("#woocommerce_checkout_page_id option:selected", data).val()
							if ("undefined" != typeof checkoutPageId && checkoutPageId.length > 0) {
								wcPages.Checkout = protocol + "//" + site + "/?p=" + checkoutPageId
							}
						}).done(function() {}))
					jQuery.when(jQuery.get(protocol + "//" + site + "/wp-admin/admin.php?page=wc-settings&tab=account",
						function(data) {
							myAccountPageId = jQuery("#woocommerce_myaccount_page_id option:selected", data).val()
							if ("undefined" != typeof myAccountPageId && myAccountPageId.length > 0) {
								wcPages["My Account"] = protocol + "//" + site + "/?p=" + myAccountPageId
							}
						}).done(function() {}))
					processResponse(wcLinks)
				}))
			} else {
				siteContent = response.content
				message = {
					"action": "update-content",
					"value": siteContent
				}
			}
			chrome.runtime.sendMessage(message, function(response) {
				return true;
			})
		})
		if (window.location.href.indexOf('edit.php?post_type=product') > 0 ) {
			match = /[0-9]*/
			jQuery('td.name.has-row-actions div.row-actions').each(function(){
				id = jQuery('span.id',jQuery(this)).html().replace(/^\D+/g, '').match(match)[0]
				jQuery(this).append('<span class="add-to-cart"> | <a href="' + protocol + '//' + site + '?add-to-cart=' + id + '">Add to cart</span>')
			})
		}
	}

	/*get the hide yoast setting, hide if appropriate*/
	chrome.storage.sync.get({
		hideYoast: true
	}, function(items) {
		if (items.hideYoast) {
			jQuery(
				".column-wpseo-score, .column-wpseo-title, #wpseo_meta, .column-wpseo-metadesc, .column-wpseo-focuskw, div.wp-pointer"
			).hide()
		}
	})

	function retrieveAjaxUrl() {
		var ret
		var scriptContent = "if ('undefined' != typeof ajaxurl) {\n";
		scriptContent += "jQuery('body').attr('tmp_ajaxurl', ajaxurl)}\n"
		scriptContent +=
			"else if ( 'undefined' != typeof woocommerce_params && 'undefined' != typeof woocommerce_params.ajax_url) {\n"
		scriptContent += "jQuery('body').attr('tmp_ajaxurl', woocommerce_params.ajax_url)\n"
		scriptContent += "}";
		var script = document.createElement('script');
		script.id = 'tmpScript';
		script.appendChild(document.createTextNode(scriptContent));
		(document.body || document.head || document.documentElement).appendChild(script);
		ret = jQuery("body").attr("tmp_ajaxurl");
		jQuery("#tmpScript").remove();
		return ret;
	}

})

chrome.runtime.onMessage.addListener(
	function(request, sender, sendResponse) {
		if ('undefined' != typeof request.message) {
			if ("toggleNotices" == request.message.action) {
				show = (noticesDisplayed) ? false : true
				jQuery(".notice, .updated, .error, .update-nag").toggle(show)
				noticesDisplayed = show
			}
			if ("hideYoast" == request.message.action) {
				show = (request.message.value) ? false : true
				jQuery(
					".column-wpseo-score, .column-wpseo-title, #wpseo_meta, .column-wpseo-metadesc, .column-wpseo-focuskw, div.wp-pointer"
				).toggle(show)
			}
			if ("whats-the-content" == request.message.action) {
				//is site content defined? if so, send that
				if ("undefined" != typeof siteContent) {
					sendResponse({
						"content": siteContent
					})
				} else { //ask the background script if it has content for this url
					message = {
						"action": "have-content",
						"value": site
					}
					chrome.runtime.sendMessage(message, function(response) {
						if (response.content) {
							content = response.content
						} else {

						}
					})
				}
				sendResponse({
					"content": siteContent
				})
			}
			if ("helper-active" == request.message.action) {
				canSwitch = (jQuery('#menu-dashboard > ul a[href$="page=woothemes-helper"]').length > 0) && (document.documentElement.lang.indexOf('en') == -1)
				sendResponse({"content": canSwitch })
			}
			if ("setLanguage" == request.message.action) {
				var expires = new Date()
   	expires.setTime(expires.getTime()+(60*1000*15))
				document.cookie = "wooninja_language=en;path=/;expires=" + expires.toUTCString() + ";"
				location.reload()
			}
			if ("resetLanguage" == request.message.action) {
				var expires = new Date()
   	expires.setTime(expires.getTime()-(60*1000*15))
				document.cookie = "wooninja_language=en;path=/;expires=" + expires.toUTCString() + ";"
				location.reload()
			}
			if ("language-overridden" == request.message.action) {
				sendResponse({"content":( document.cookie.indexOf("wooninja_language") > -1 ) ? true : false})
			}
		}
	})

function processResponse(wcLinks) {
	var html
	var wcMenu = $("#toplevel_page_woocommerce ul.wp-submenu")
	if (wcMenu.length < 1) {
		$.when($.get(protocol + "//" + site + "/wp-admin/", function(data) {})).done(function(data) {
			wcMenu = $("#toplevel_page_woocommerce ul.wp-submenu", data)
		})
	}
	subLinks = $('li a', wcLinks)
	var linkSet
	var numLoops = subLinks.length
	$(subLinks).each(function(index) {
		var sub = $(this)
		var theLinks = {}
		var href = $(this).attr('href')
		$.when($.get(href, function(subData) {
			underLinks[index] = $('ul.subsubsub', subData)
		}).done(function() {
			if (underLinks.length >= numLoops) {
				html = $(wcLinks).html().split('<li>')
				for (key in underLinks) {
					htmlKey = parseInt(key) + 1
					if ("undefined" != typeof underLinks[key][0]) {
						/*
						TODO: there should be some way to identify what these links point to
						and get their English titles and output that as the title in the
						case of working on a non-English site
						 */
						html[htmlKey] += "<li>" + underLinks[key][0].outerHTML + "</li>"
					}
				}
				html[0] = '<li class="no-indent">' + html[0] + "</li>"
				html = '<ul>' + html.join('<li>') + '</ul>'
				html += "<ul>"
				$("a", wcMenu).each(function() {
					href = protocol + "//" + site + "/wp-admin/" + $(this).attr("href")
					text = $(this).text()
					html += '<li><a href="' + href + '" tabindex="-1">' + text + '</a></li>'
				})
				html += "</ul>"
				html += "<hr>Frontend:<br><ul>"
				for (index in wcPages) {
					if (null !== wcPages[index]) {
						html += "<li><a href=" + wcPages[index] + ">" + index + "</a></li>"
					} else {
						html += "<li>" + index + " is undefined</li>"
					}
				}
				html += "</ul>"
				siteContent = html
				message = {
					"action": "set-content",
					"value": {
						"site": site,
						"content": siteContent
					}
				}
				chrome.runtime.sendMessage(message, function(response) {
					return true;
				})
			}
		}))
	})
}


options.js
/**
 * License: GPLv2 http://www.gnu.org/licenses/gpl-2.0.html
 * Copyright (C) 2015  Will Brubaker
	* This program is free software; you can redistribute it and/or
	* modify it under the terms of the GNU General Public License
	* as published by the Free Software Foundation; either version 2
	* of the License, or (at your option) any later version.
	*
	* This program is distributed in the hope that it will be useful,
	* but WITHOUT ANY WARRANTY; without even the implied warranty of
	* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	* GNU General Public License for more details.
	*
	* You should have received a copy of the GNU General Public License
	* along with this program; if not, write to the Free Software
	* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// Saves options to chrome.storage.sync.
function save_options() {
	var hideYoast = document.getElementById('hide-yoast').checked
	chrome.storage.sync.set({
		hideYoast: hideYoast,
	}, function() {
		// Update status to let user know options were saved.
		var status = document.getElementById('status')
		status.textContent = 'Options saved.'
		setTimeout(function() {
			status.textContent = '';
		}, 750);
	});
}

// Restores select box and checkbox state using the preferences
// stored in chrome.storage.
function restore_options() {
	chrome.storage.sync.get({
		hideYoast: true,
	}, function(items) {
		document.getElementById('hide-yoast').checked = items.hideYoast
	});
}
document.addEventListener('DOMContentLoaded', restore_options);
document.getElementById('save').addEventListener('click',
	save_options);


background.js
/**
 * License: GPLv2 http://www.gnu.org/licenses/gpl-2.0.html
 * Copyright (C) 2015  Will Brubaker
	* This program is free software; you can redistribute it and/or
	* modify it under the terms of the GNU General Public License
	* as published by the Free Software Foundation; either version 2
	* of the License, or (at your option) any later version.
	*
	* This program is distributed in the hope that it will be useful,
	* but WITHOUT ANY WARRANTY; without even the implied warranty of
	* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	* GNU General Public License for more details.
	*
	* You should have received a copy of the GNU General Public License
	* along with this program; if not, write to the Free Software
	* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
var sitesContent = {}
chrome.runtime.onMessage.addListener(
	function(request, sender, sendResponse) {
		if ("undefined" != typeof request.action) {
			if ("have-content" == request.action) {
				//check if sitesContent request.value.site is defined if it is, send content
				//to the panel, if not send false
				site = request.value.site
				content = ( "undefined" != typeof sitesContent[site] ) ? sitesContent[site].content : false
				sendResponse({
					"content": content
				})
			}
			if ("set-content" == request.action) {
				sitesContent[request.value.site] = {"content": request.value.content}
			}
		}
	})

panel.js
/**
 * License: GPLv2 http://www.gnu.org/licenses/gpl-2.0.html
 * Copyright (C) 2015  Will Brubaker
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
restore_options()
var tabContent

/* hides or shows WordPress admin notices (for screenshots or whatever) */
$("#toggle-notices").on("click", function() {
	chrome.tabs.query({
		active: true,
		currentWindow: true
	}, function(tabs) {
		message = {
			"action": "toggleNotices",
			"value": true
		}
		chrome.tabs.sendMessage(tabs[0].id, {
			message: message
		}, function(response) {
			return true
		})
	})
	window.close()
})

$("#engbutton").on("click",function(){
	chrome.tabs.query({
		active: true,
		currentWindow: true
	}, function(tabs) {
		message = {
			"action": "setLanguage",
			"value": true
		}
		chrome.tabs.sendMessage(tabs[0].id, {
			message: message
		}, function(response) {
			return true
		})
	})
	window.close()
})

$("#reset").on("click",function(){
	chrome.tabs.query({
		active: true,
		currentWindow: true
	}, function(tabs) {
		message = {
			"action": "resetLanguage",
			"value": true
		}
		chrome.tabs.sendMessage(tabs[0].id, {
			message: message
		}, function(response) {
			return true
		})
	})
	window.close()
})

/* updates the extension settings from the panel */
jQuery("#hide-yoast").on("change", function() {
	hideYoast =
		chrome.storage.sync.set({
			hideYoast: jQuery(this).prop("checked")
		})
		/* send a message to the content script to toggle as appropriate */
	chrome.tabs.query({
		active: true,
		currentWindow: true
	}, function(tabs) {
		message = {
			"action": "hideYoast",
			"value": document.getElementById('hide-yoast').checked
		}
		chrome.tabs.sendMessage(tabs[0].id, {
			message: message
		})
	})
	window.close()
})

/* gets the "hide yoast" setting from the options panel and checks/unchecks the box on the panel */
function restore_options() {
	chrome.storage.sync.get({
		hideYoast: true
	}, function(items) {
		document.getElementById('hide-yoast').checked = items.hideYoast;
		if (true == items.hideYoast) {
			chrome.tabs.query({
				active: true,
				currentWindow: true
			}, function(tabs) {
				message = {
					"action": "hideYoast",
					"value": true
				}
				chrome.tabs.sendMessage(tabs[0].id, {
					message: message
				})
			})
		}
	})
}

chrome.runtime.onMessage.addListener(
	function(request, sender, sendResponse) {
		if ("undefined" != typeof request && "undefined" != typeof request.action) {
			if ("update-content" == request.action) {
				tabContent = request.value
				$("#wc-links").empty().append(tabContent)
			}
		}
		return true
	})

jQuery(document).ready(function() {
	//ask the content script for the tab content
	message = {
		"message": {
			"action": "whats-the-content"
		}
	}
	chrome.tabs.query({
		active: true,
		currentWindow: true
	}, function(tabs) {
		chrome.tabs.sendMessage(tabs[0].id, message, function(response) {
			if ("undefined" != typeof response && "undefined" != typeof response.content) { //if there is content, put the links in the thing, man
				$("#wc-links").empty().append(response.content)
				jQuery("a").each(function() {
					jQuery(this).qtip({
						content: {
							text: jQuery(this).attr("href")
						},
						position: {
							my: "bottom left",
							at: "bottom left",
							target: jQuery(window)
						},
						style: {
							classes: "qtip-tipsy"
						},
						show: {
							delay: 300
						}
					})
				})
				jQuery("a").on("click", function() {
					href = jQuery(this).attr("href")
					chrome.tabs.update(tabs[0].id, {
						url: href
					})
					window.close()
				})
			}
		})
		message = {
			"message": {
				"action": "helper-active"
			}
		}
		chrome.tabs.sendMessage(tabs[0].id, message, function(response) {
				if ("undefined" != typeof response && "undefined" != typeof response.content) {
					jQuery('#engbutton').toggle(response.content)
				}
		})
		message = {
			"message": {
				"action": "language-overridden"
			}
		}
		chrome.tabs.sendMessage(tabs[0].id, message, function(response) {
				if ("undefined" != typeof response && "undefined" != typeof response.content) {
					jQuery('#resetlang').toggle(response.content)
				}
		})
	})
})
